### DISCUSSION TITLE
Unattached docs not rendering theme

### DISCUSSION THREAD

**Discussion by atomicrobokid** [storybookjs/storybook#28495] [Help]
### Summary

Hello, hoping someone can help me with an Angular based storybook implementation (v8).

I've got my `preview.ts` set up with the below config to add a data attribute (which renders on the html tag to set the theme)
```
  decorators: [
    withThemeByDataAttribute<AngularRenderer>({
      attributeName: 'data-theme',
      parentSelector: 'html',
      defaultTheme: 'light',
      themes: {
        light: 'light',
        dark: 'dark'
      },
    }),
  ],
    parameters: {
      themes: {
        clearable: false,
        default: 'light',
        target: 'html',
        property: 'data-theme',
        options: [
          { name: 'light', value: 'light', color: '#f5f6f7' },
          { name: 'dark', value: 'dark', color: '#443443' }
        ],
      },
    }
```

This works perfectly for Stories, and for `.mdx` files that contain a story, but for vanilla unattached `.mdx` files with just static content, this does not work, the prop on the html tag does not update.

A few other oddities i've noticed. The api docs for `storybook-addon-themes` say : "target (string - optional): Target element selected with document.querySelector() to which classes are applied. Defaults to body, root if classes should be applied to documentElement." - however no matter what i set the target to, the attribute _always_ renders on the html tag.

Another one, is if i visit a story and switch the theme (which works) and then go to an unattached mdx page, the html data attribute leaks onto the new page, i.e the theme i've switched to renders, however it still cannot be switched.

Any help is super appreciated, tried all sorts of config and no joy so far!

*edit* - Looks to be similar to [#26625](https://github.com/storybookjs/storybook/issues/26625) and the last comment in [#25183](https://github.com/storybookjs/storybook/discussions/25183)

**Comment by atomicrobokid** [comment]
Ok after quite a bit of fiddling, i've got it working.

So for anyone else in the future looking for a quick and dirty way to style vanilla mdx (especially in an Angular app, as the docs for writing a custom docs wrapper just didn't work) 

Assuming you have `@storybook/addon-themes` set up like below (this is a decorator in your `preview.ts`)

```
  withThemeByDataAttribute<AngularRenderer>({
    attributeName: 'data-theme',
    parentSelector: 'html',
    defaultTheme: 'light',
    themes: {
      light: 'light',
      dark: 'dark'
    },
  })
```

Next add this function to your docs property in `preview.ts`
```
    docs: {
      container: (props: any) => {
        const el = document.querySelector("html");
        el!.dataset['theme'] = props?.context.store.globals.globals.theme;
        return createElement(DocsContainer, props);
      },
    },
```

Appreciate getting it from the store is likely to break at _some_ point, but for now, it works and can't find any other Angular friendly solutions. Would be nice to have this working for unattached mdx out of the box so you can use your design systems themes without a load of extra config

**Reply by declanchiu** [reply]
Thank you very much for providing a solution, can I ask a question, where does createElement come from?

**Reply by atomicrobokid** [reply]
@declanchiu - no problem, that comes from the react package which is a dependency installed by `addon-docs` 

`import { createElement } from 'react';`

**Reply by atomicrobokid** [reply]
Another small but quite important update, if you're using any storybook blocks such as `ColorPalette` etc and want that theme to also change alongside adding some theming data-attribute:

```
      theme: themes.light,
      container: (props: any) => {
        const el = document.querySelector('html');
        el!.dataset['theme'] = props?.context.store.globals.globals.theme;
        const theme = props?.context.store.globals.globals.theme === 'dark' ? themes.dark: themes.light
        props.theme = theme;
        return createElement(DocsContainer, props);
      },
```

**Reply by declanchiu** [reply]
Thanks a lot.

**Reply by declanchiu** [reply]
It looks like the props.theme = theme behavior is not allowed in react
![image](https://github.com/user-attachments/assets/a956f46f-1fcd-4bab-9edc-0ae5c8928709)

**Reply by declanchiu** [reply]
Thanks, that's my problem. I could have used a shallow copy.

![image](https://github.com/user-attachments/assets/f3274fa5-414d-4a01-911f-9c89aa5e6763)

**Reply by arnoldsandoval** [reply]
I got this working in storybook 8 with a minor modification on top of @atomicrobokid and @declanchiu's solution.

inside of `parameters.docs`
```
  parameters: {
     // ...
    docs: {
      container: (props: any) => {
        const el = document.querySelector("html");
        const theme =
          props?.context.store.userGlobals.globals.theme === "dark"
            ? themes.dark
            : themes.light;
        el!.dataset["theme"] = props?.context.store.userGlobals.globals.theme;
        const newProps = { ...props, theme };
        return React.createElement(DocsContainer, newProps);
      },
    },
  },
```

Example: https://ui.otti.dev/?path=/docs/charts-stackedhorizontalbarpair--documentation&globals=theme:dark

**Reply by reduckted** [reply]
> I got this working in storybook 8 with a minor modification on top of @atomicrobokid and @declanchiu's solution.

Nice. 🎉  For how I'm setting the theme, the key was to use `store.userGlobals` instead of `store.globals` to get the theme being used.

**Reply by duncan-s-white** [reply]
Thanks so much guys!! This helped fix the same issue with a nextjs storybook 8 installation.

I was slightly confused where a couple of things were coming from, as I wasn't getting any import suggestions in VSCode, so I include the imports I needed if it helps anyone else:

```
import { themes } from '@storybook/theming';
import { DocsContainer } from '@storybook/blocks';
```

**Comment by toppsdown** [comment]
I'm trying to use the workarounds described here, but they're causing this [error](https://github.com/storybookjs/storybook/issues/26820):

TypeError: Cannot read properties of undefined (reading 'fonts')

Looking at the source, it appears to be related to the Code component: // src/components/components/typography/elements/Code.tsx

There's a proposed workaround [here](https://github.com/storybookjs/storybook/issues/29313#issuecomment-2403915357) but it's not working for me.
